distributed/scheduler.py

```diff
@@ -2218,6 +2218,22 @@ def transition_waiting_processing(self, key: str, stimulus_id: str) -> RecsMsgs:
                 if not (ws := self.decide_worker_rootish_queuing_enabled()):
                     return {ts.key: "queued"}, {}, {}
         else:
+            if not math.isinf(self.WORKER_SATURATION):
+                # Queuing can break priority ordering, e.g. when there are
+                # worker restrictions.
+                # We need to check here if there is a more important queued task
+                # and send the currently transitioning task back in the line to
+                # avoid breadth first search
+                # See also https://github.com/dask/distributed/issues/7496
+                slots_available = sum(
+                    _task_slots_available(ws, self.WORKER_SATURATION)
+                    for ws in self.idle_task_count
+                )
+
+                for qts in self.queued.peekn(slots_available):
+                    if qts.priority < ts.priority:
+                        return {ts.key: "queued"}, {}, {}
```

Comment:
Just to be clear, we're now intentionally putting non-rootish tasks which potentially have restrictions into the queue? That will fail this validation at least:
https://github.com/dask/distributed/blob/70abff08f3cb715f6641fc4b8ede1d9f0c62ad91/distributed/scheduler.py#L5010-L5012